src/rev.md

```diff
@@ -19,58 +19,63 @@ Both the inplace and "normal" variant return the gradient. The difference is tha
 
 ```rust
 #[autodiff(df, Reverse, Active, Active, Active)]
 fn f(x: f32, y: f32) -> f32 {
-  x*x + 3.0 * y
+  x * x + 3.0 * y
 }
 ```
 
 Enzyme actually generates the code on LLVM-IR level, but Rust is nicer to read, so I will pretend we would generate a Rust implementation:
 
 ```rust
 fn f(x: f32, y: f32) -> f32 {
-  x*x + 3.0 * y
+  x * x + 3.0 * y
 }
 fn df(x: f32, y: f32) -> (f32, f32, f32) {
-  (2.0 * x, 3.0, x*x + 3.0 * y)
+  let d_dx = 2.0 * x;
+  let d_dy = 3.0;
+  let f = x * x + 3.0 * y;
+  (d_dx, d_dy, f)
 }
 ```
 
-$x*x$ becomes $2.0*x$, while $3.0 * y$ becomes $3.0$. The last argument is our original return value. However, we don't always pass things by value, so let's make sure we have a sensible solution:
+Note that the last entry in the result tuple contains the original return value. However, we don't always pass things by value, so let's make sure we have a sensible solution:
 
 ```rust
 #[autodiff(df, Reverse, Active, Duplicated, Active)]
 fn f(x: f32, y: &f32) -> f32 {
-  x*x + 3.0 * y
+  x * x + 3.0 * y
 }
 ```
 
-(pay attention to y).
+(pay attention to `y`).
 
 ```rust
 fn f(x: f32, y: f32) -> f32 {
-  x*x + 3.0 * y
+  x * x + 3.0 * y
 }
-fn df(x: f32, y: &f32, dy: &mut f32) -> (f32, f32) {
-  dy += 3.0;
-  (2.0 * x, x*x + 3.0 * y)
+fn df(x: f32, y: &f32, d_dy: &mut f32) -> (f32, f32) {
+  let d_dx = 2.0 * x;
+  *d_dy += 3.0;
+  let f = x * x + 3.0 * y
+  (d_dx, f)
 }
 ```
 
-In the case of references (or pointers) we do expect the user to create `dy`.
+In the case of references (or pointers) we do expect the user to create `d_dy`.
 
-We could obviously zero-initialize a float for the user, but let's assume the constructor is complex due to involving a double-linked-list or ffi, so we can't guarantee that on the compiler side. As an alternative motivation, imagine that we call `df` 5 times in a loop. It is clear that in this case the accumulated gradients should be 5 times higher too, which won't happen if we set `dy = 3.0` each time, instead of using `+=`. Yet another reason would be higher-order derivatives (todo: just refer to literature?).
+We could obviously zero-initialize a float for the user, but let's assume the constructor is complex due to involving a double-linked-list or ffi, so we can't guarantee that on the compiler side. As an alternative motivation, imagine that we call `df` 5 times in a loop. It is clear that in this case the accumulated gradients should be 5 times higher too, which won't happen if we set `d_dy = 3.0` each time, instead of using `+=`. Yet another reason would be higher-order derivatives (todo: just refer to literature?).
 
 Now that we got back from this rabbit hole, let's go wild and train a neural network on our local national lab server:
 
 ```rust
 #[autodiff(backprop, Reverse, Duplicated, Duplicated, Active)]
-fn inference(images: &[f32], weights: &[f32]) -> f32 {
+fn training_loss(images: &[f32], weights: &[f32]) -> f32 {
```

Comment:
I just can't stomach "inference" for a model that predicts observables. (Before the self-aggrandizing ML terminology entered the lexicon with the opposite meaning, there was consensus that the word refers to estimating latent quantities.) But anyway, this is a loss function, not just evaluating the forward `predict` (that ML people call inference). Technically, we would have labels or known quantities also as input to the loss function, but we can gloss over that.

Reply:
fair enough